File: modules/core/src/main/java/com/panayotis/jubler/subs/style/gui/AlphaColor.java
/*
 * (c) 2005-2023 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package  com.panayotis.jubler.subs.style.gui;

import java.awt.Color;

public class AlphaColor extends Color {

    private int alpha;

    public AlphaColor(int color) {
        super(color & 0xffffff);
        alpha = (color >> 24) & 0xff;
    }

    public AlphaColor(Color c, int alpha) {
        super(c.getRGB());
        this.alpha = alpha;
    }

    public AlphaColor(String color) {
        super(Integer.parseInt(color.substring((color.length() > 6) ? color.length() - 6 : 0, color.length()), 16));
        alpha = (color.length() > 6) ? Integer.parseInt(color.substring(0, color.length() - 6), 16) : 0;
    }

    public AlphaColor(AlphaColor c) {
        this(c, c.alpha);
    }

    public Color getMixed(Color other, int newalpha) {
        float a = newalpha / 255f;
        float na = 1 - a;
        return new Color(calcColor(getRed(), other.getRed(), a, na),
                calcColor(getGreen(), other.getGreen(), a, na),
                calcColor(getBlue(), other.getBlue(), a, na));
    }

    public Color getMixed(Color other) {
        return getMixed(other, getAlpha());
    }

    private int calcColor(int col1, int col2, float a, float na) {
        int ret = (int) (col1 * a + col2 * na);
        if (ret < 0)
            ret = 0;
        if (ret > 255)
            ret = 255;
        return ret;
    }

    public Color getAColor() {
        return new Color(getRed(), getGreen(), getBlue(), alpha);
    }

    public int getAlpha() {
        return alpha;
    }

    public int getARGB() {
        return (getRGB() & 0xffffff) | (alpha << 24);
    }

    public String toString() {
        String res = Integer.toHexString(getRGB() & 0xffffff);
        String alp = Integer.toHexString(alpha);
        return "00".substring(0, 2 - alp.length()) + alp + "000000".substring(0, 6 - res.length()) + res;
    }

    public boolean equals(Object other) {
        if (other == null)
            return false;
        if (other instanceof AlphaColor) {
            if (((AlphaColor) other).alpha != alpha)
                return false;
            return super.equals(other);
        }
        return false;
    }
}

File: src/core/style/AlphaColor.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/style/AlphaColor.h"

AlphaColor::AlphaColor(const QString &text) {
    const int len = text.length();
    const int split = len > 6 ? len - 6 : 0;
    bool ok = false;
    rgb_ = text.mid(split).toUInt(&ok, 16) & 0xffffff;
    if (!ok)
        rgb_ = 0;  // Java would throw NumberFormatException; callers pass valid data
    alpha_ = len > 6 ? int(text.left(split).toUInt(&ok, 16) & 0xff) : 0;
}

namespace {
int calc(int c1, int c2, float a, float na) {
    int ret = int(c1 * a + c2 * na);
    return ret < 0 ? 0 : (ret > 255 ? 255 : ret);
}
}  // namespace

QColor AlphaColor::mixed(const QColor &other, int newAlpha) const {
    const float a = newAlpha / 255.0f;
    const float na = 1 - a;
    return QColor(calc(red(), other.red(), a, na), calc(green(), other.green(), a, na),
                  calc(blue(), other.blue(), a, na));
}

QString AlphaColor::toString() const {
    return QStringLiteral("%1%2")
        .arg(alpha_, 2, 16, QLatin1Char('0'))
        .arg(rgb_, 6, 16, QLatin1Char('0'));
}
